pkg/restclient/roundtrippers.go

```diff
@@ -0,0 +1,88 @@
+package restclient
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// debugRoundTripper is a middleware which debugs each outgoing request.
+// It will dump information about request passing through it.
+type debugRoundTripper struct {
+	rt http.RoundTripper
+}
+
+func newDebugRoundTripper(rt http.RoundTripper) http.RoundTripper {
+	return &debugRoundTripper{rt}
+}
+
+// RoundTrip dumps the request and response of a single http transaction.
+//
+// It will dump each outgoing request in its HTTP/1.x wire representation and
+// also It will dump the received response in its HTTP/1.x wire representation.
+func (d *debugRoundTripper) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	reqCtx := newRequestCtx(req)
+	logrus.Debug(reqCtx.curl())
+
+	if reqDump, err := httputil.DumpRequestOut(req, true); err == nil {
+		logrus.Debugf("Trying %s... \n%s", reqCtx.reqURL.Host, reqCtx.dump(">", reqDump))
```

Comment:
Try not to use `Debugf()` or other format string log methods. Instead try to use structured logging like below,
```
log.WithFields(log.Fields{
"host": reqCtx.reqURL.Host,
"httprequest": dump(),
}.Debug("Sending request")
```
Change the rest of the messages to be similar to the above.

One thing though, the `dump()` function will break a single log across multiple lines. This make logs harder to parse with the current format (if we used configured logrus to log in json, say for example, it wouldn't be a problem). I'm sorta okay with this happening as this is a debug log, but would have preferred if the log was all in one line.  @prashanthpai do you have any opinions on this?

Reply:
since request Dump contains new line ('\n'), if we pass request dump data in log.Fields, then instead of printing in new line it will simply print '\n' in log fields.